shell/platform/darwin/ios/framework/Headers/FlutterEngine.h

```diff
@@ -99,10 +105,10 @@ FLUTTER_EXPORT
  * immediately.
  *
  * @param entrypoint The name of a top-level function from the same Dart
- *   library that contains the app's main() function.  If this is nil, it will
- *   default to `main()`.  If it is not the app's main() function, that function
- *   must be decorated with `@pragma(vm:entry-point)` to ensure the method is not
- *   tree-shaken by the Dart compiler.
+ *   library that contains the app's main() function.  If this is FlutterDefaultDartEntrypoint (or
+ *   nil) it will default to `main()`.  If it is not the app's main() function, that function must
+ *   be decorated with `@pragma(vm:entry-point)` to ensure the method is not tree-shaken by the Dart
+ *   compiler.
  * @return YES if the call succeeds in creating and running a Flutter Engine instance; NO otherwise.
  */
 - (BOOL)runWithEntrypoint:(nullable NSString*)entrypoint;
```

Comment:
I'm not familiar with how things are packaged at the binary level in iOS, but do we not need a configurable bundle path along with the function name? The entrypoint on Android has both pieces in a data structure called DartEntrypoint...

Reply:
https://github.com/flutter/engine/blob/cd0330e6e0ce68949e1b8cbae6e66eed3e74f426/shell/platform/darwin/ios/framework/Headers/FlutterEngine.h#L63 . You specify the bundle path when you create the FlutterEngine.

Reply:
Is there a reason that's separated? Is the goal of this change to match Android? If so, shouldn't that involve pairing the entrypoint with the bundle path? The entrypoint only exists in a specific bundle, so if we think it's legitimate to lazily choose the entrypoint method, then it seems like you should be able to lazily choose the bundle path, right?

Reply:
I think that's a legitimate point.  What you are proposing is a breaking change though.  The idea with this PR is it gets us closer to Android without breaking.  We can see what @xster thinks.  I'd say that could be a different PR though.

Reply:
We can see what @xster says, but seems like we could introduce a new constructor and a new dart entrypoint setter method, then soft deprecate the existing ones. That wouldn't be a breaking change, but it would move us forward for the future.

Reply:
Adding a new init could make things a bit confusing since what would it mean if you specified "foo" as your entrypoint in the init but specified "bar" as your entrypoint in the run function?  We would have to deprecate both the old init and run functions.  While they co-existed it would be a bit weird.

The issue doesn't seem to reference the java code.  Could you link it here what you are referring to?  This is what the java documentation says currently:
```
    // Create and pre-warm a FlutterEngine.
    FlutterEngine flutterEngine = new FlutterEngine(context);
    flutterEngine
      .getDartExecutor()
      .executeDartEntrypoint(DartEntrypoint.createDefault());
 ```
Are you talking about `FlutterRunArguments`?

Reply:
That DartEntrypoint contains 2 fields: an entrypoint function and a bundle path:

https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/engine/dart/DartExecutor.java#L277

Reply:
ya, we're in kind of a pickle. The Android engine constructor doesn't ask for anything an everything's provided by the dart executor's execute. 

The iOS engine asks for half of the stuff during init and the other half during run. 

I think the ideal solution is probably:

- Deprecate the `initWithName:(NSString*)labelPrefix project:(FlutterDartProject*)project` init. If you give it a project, we'll just ignore it. 
- Let the `launchEngine:(NSString*)entrypoint libraryURI:(NSString*)libraryOrNil` also have a `withProject:(FlutterDartProject*)project` to fulfill this API.
- Let there also be a launchDefaultEngine or some such so you can just `[[[FlutterEngine alloc] initWithName:@"blah"] launchDefaultEngine]` so that the most default engine can be launched without params or nils like Android. 

It looks like _dartProject is only ever used in the launchEngine method in the FlutterEngine so moving it implementation wise shouldn't alter too much. 

The break would be ok (but still needs announcing) since the only valid use of that API is add-to-app currently and we never shipped add-to-app yet.

Reply:
Hmm, I'll look into making this a bit deeper.  If you want the API's to be closer though, shouldn't we introduce a class called FlutterDartEntrypoint in objc and make a method on `FlutterEngine` `runWithDartEntrypoint:`?

Reply:
I'm unable to add the bundle to the run arguments.  This would negatively affect our app launch time for full flutter apps since the engine is created and `launch` is only called after `viewWillAppear:`

There are still a few things I can do to make working with the FlutterEngine nicer though.  I'll do everything but move the bundle to run/launch arguments.